tests/phpunit/Civi/Core/FormatTest.php

```diff
@@ -0,0 +1,228 @@
+        'money_number' => '1 234,56',
+        'number' => '1 234,56',
+        'money_long' => '1 234,56 €',
+        'money_number_long' => '1 234,56',
+      ],
+    ];
+    $cases['fr_FR_EUR_long'] = [
+      [
+        'amount' => '1234.56700',
+        'locale' => 'fr_FR',
+        'currency' => 'EUR',
+        'money' => '1 234,57 €',
+        'money_number' => '1 234,57',
+        'number' => '1 234,567',
+        'money_long' => '1 234,567 €',
+        'money_number_long' => '1 234,567',
+      ],
+    ];
+    $cases['fr_FR_EUR_pad'] = [
+      [
+        'amount' => '1234.50',
+        'locale' => 'fr_FR',
+        'currency' => 'EUR',
+        'money' => '1 234,50 €',
+        'money_number' => '1 234,50',
+        'number' => '1 234,5',
+        'money_long' => '1 234,50 €',
+        'money_number_long' => '1 234,50',
+      ],
+    ];
+    $cases['ar_AE_KWD'] = [
+      [
+        'amount' => '1234.56',
+        'locale' => 'ar_AE',
+        'currency' => 'KWD',
+        'money' => '١٬٢٣٤٫٥٦٠ د.ك.‏',
+        'money_number' => '١٬٢٣٤٫٥٦٠',
+        'number' => '١٬٢٣٤٫٥٦',
+        'money_long' => '١٬٢٣٤٫٥٦٠ د.ك.‏',
+        'money_number_long' => '١٬٢٣٤٫٥٦٠',
+      ],
+    ];
+    $cases['ar_AE_KWD_long'] = [
+      [
+        'amount' => '1234.56710',
+        'locale' => 'ar_AE',
+        'currency' => 'KWD',
+        'money' => '١٬٢٣٤٫٥٦٧ د.ك.‏',
+        'money_number' => '١٬٢٣٤٫٥٦٧',
+        'number' => '١٬٢٣٤٫٥٦٧١',
+        'money_long' => '١٬٢٣٤٫٥٦٧١ د.ك.‏',
+        'money_number_long' => '١٬٢٣٤٫٥٦٧١',
+      ],
+    ];
+    $cases['ar_AE_KWD_pad'] = [
+      [
+        'amount' => '1234.56',
+        'locale' => 'ar_AE',
+        'currency' => 'KWD',
+        'money' => '١٬٢٣٤٫٥٦٠ د.ك.‏',
+        'money_number' => '١٬٢٣٤٫٥٦٠',
+        'number' => '١٬٢٣٤٫٥٦',
+        'money_long' => '١٬٢٣٤٫٥٦٠ د.ك.‏',
+        'money_number_long' => '١٬٢٣٤٫٥٦٠',
+      ],
+    ];
+    $cases['en_US_KWD'] = [
+      [
+        'amount' => '1234.56',
+        'locale' => 'fr_FR',
+        'currency' => 'KWD',
+        'money' => '1 234,560 KWD',
+        'money_number' => '1 234,560',
+        'number' => '1 234,56',
+        'money_long' => '1 234,560 KWD',
+        'money_number_long' => '1 234,560',
+      ],
+    ];
+    $cases['en_US_KWD_long'] = [
+      [
+        'amount' => '1234.5678000',
+        'locale' => 'fr_FR',
+        'currency' => 'KWD',
+        'money' => '1 234,568 KWD',
+        'money_number' => '1 234,568',
+        'number' => '1 234,5678',
+        'money_long' => '1 234,5678 KWD',
+        'money_number_long' => '1 234,5678',
+      ],
+    ];
+    $cases['en_US_KWD_pad'] = [
+      [
+        'amount' => '1234.5',
+        'locale' => 'fr_FR',
+        'currency' => 'KWD',
+        'money' => '1 234,500 KWD',
+        'money_number' => '1 234,500',
+        'number' => '1 234,5',
+        'money_long' => '1 234,500 KWD',
+        'money_number_long' => '1 234,500',
+      ],
+    ];
+    return $cases;
```

Comment:
I figured en_IN / INR would be interesting to try (eg the comma rules are different). This is what currently passes on my system:

```suggestion
    $cases['en_IN_INR'] = [
      [
        'amount' => '123456789',
        'locale' => 'en_IN',
        'currency' => 'INR',
        'money' => '₹ 12,34,56,789.00',
        'money_number' => '12,34,56,789.00',
        'number' => '12,34,56,789',
        'money_long' => '₹ 12,34,56,789.00',
        'money_number_long' => '12,34,56,789.00',
      ],
    ];

    return $cases;
```

It handled the 22,22,333 commas (yay) - but the decimals seem a little suspicious to me. My recollection is that prices are usually whole rupees, and https://en.wikipedia.org/wiki/Indian_paisa says "As of 2019, all paisa coins have been demonetised." So I'm not sure if that's a subjective policy issue with the currency database (*"the change was so recent... we still need to present 50p transactions!"*) or if that's a technical issue ("the formatter doesn't handle 0-decimals").

Reply:
The decimal nuance comes from the brick money library - I think they talk about Japanese having no decimal places - but I do think it relies on a list within that library

Reply:
Yeah, upstream list describes INR as 2-decimal-digits. It only provides one length (eg `getDefaultFractionDigits()`) whereas `NumberFormatter` allows a distinction between min/max fraction digits. Maybe this is a use-case for min/max digits? (ie "Most data is `.00` and that should be dropped,  but if your data says `.50` then show it".) But obviously, someone (somewhere) would need to hardcode an opinion of `INR, 0-2 digits`, which disagrees with our current data.

So there might be some room for improvement in that, but I think the new API (`Civi::format()->money(...)`) should still look the same. It shouldn't block refactoring/improvement of the interface.

Reply:
@totten yeah - I think the upstream library has enough use we can rely on someone who has first-hand knowledge of a currency patching it over time